### PR TITLE
WIP(routing): URL-template-based routing

### PR DESCRIPTION
**Pull request by ewinslow**
Routing is too manual of a process in 1.x. We can streamline linking up of URLs to `resources/*` views significantly by allowing use of URL templates.

Refs #4820 

It would be cool and all to get this in for 2.0, but wouldn't be surprised if it slips 

Essentially how it works is that you define a routes.php file in the root of your plugin that has a map from URL patterns to resources/\* views:

``` php
// routes.php
return [
  '/some/url' => 'hello/world',

  // Example with a parameterized segment
  '/some/{input}' => 'hello/world',
];
```

To make "/some/url" do anything, you have to define a view at `views/default/resources/hello/world.php`. Whatever is echoed from that resource is what is returned to the requesting client.

### CODE REVIEW

**Comment by hypeJunction** [comment]
:+1

**Comment by ewinslow** [comment]
Should we also allow nested URL structure to avoid repetetive urls?

```php
return [
  '/blog' => [
    '' => 'blog/all',
    '/all' => 'blog/all',
    '/owner/{username}' => 'blog/owner',
  ],
];
```

**Comment by mrclay** [comment]
Nesting seems low value for the added complexity

**Comment by mrclay** [comment]
Flat is more readable to me as well

File: engine/lib/admin.php [hunk 1]
```diff
@@ -1,745 +0,0 @@
...
-				$widget->move($column, $position);
-			}
-		}
-	}
-	elgg_set_ignore_access(false);
-}
-
-return function(\Elgg\EventsService $events, \Elgg\HooksRegistrationService $hooks) {
-	$events->registerHandler('init', 'system', '_elgg_admin_init');
-};
```
Comment (juho-jaakkola) [inline engine/lib/admin.php]: I assume contents of this file have been removed accidentally?

**Comment by juho-jaakkola** [comment]
I also prefer flat structure

**Comment by ewinslow** [comment]
Sounds good! Flat it is then!

**Comment by ewinslow** [comment]
I'm intentionally ignoring this until after 2.0.0 ships since I think we can get it in without breaking changes.

**Comment by hypeJunction** [comment]
@ewinslow I think I will eventually hijack and finish this. Did you implement the route matching logic or what were your plans for it (Symfony?)?

**Comment by hypeJunction** [comment]
Worth considering what community plugins are doing in this regard: https://github.com/Elgg/community_plugins/blob/master/lib/page_handlers.php#L10

**Comment by hypeJunction** [comment]
I am hijacking this.

Based on lessons learnt in #10289, all our resource views do is assemble a configure arrays for layouts, and then inject them into the page view. Instead, I want a hook based approach that eliminates 5 different steps needed to render a page. With the following, we will head towards having normal controllers + the HTTP reponses will make this a flexible and easy to extend approach.

- Use `elgg-plugin.php` to configure routes
- Add API to register/override routes
- Kill `elgg_register_page_handler()`
- Kill `elgg_view_resource()`
- Implement a hook based approach or `\Elgg\Route` interface for:
     - `getPageOwner()`
     - `getPageEntity()`
     - `getHead()`  (`getMetatags()`)
     - `getBreadcrumbs()`
     - `getFilter()`
     - `getLayout()`
     - `getPageShell()`
     - `getTitle()`
     - `getContent()`
          - `get_entity()` + `elgg_view_entity()` = `elgg_list_entities(['guid' => $guid)`
          - it would be ideal if we could work with ege* options instead of rendered lists, as it would help
            with non-default viewtypes and other use cases (RESTful/WS services)
     - `getSidebar()`
     - `getSidebarAlt()`

**Comment by hypeJunction** [comment]
FYI, I don't think we are ready for full fetched controllers yet. I just want to simplify the steps needed between `route` and page view.

**Comment by hypeJunction** [comment]
See #10319 for some diarrhea of ideas.

**Comment by hypeJunction** [comment]
Replaced by #10698

**Pull request closed without merging**